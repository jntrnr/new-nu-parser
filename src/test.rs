use crate::resolver::Resolver;
use crate::{compiler::Compiler, parser::Parser};
use std::path::Path;

fn evaluate_example(fname: &Path) -> String {
    let mut compiler = Compiler::new();
    let contents = std::fs::read(fname).expect("We only run tests found by glob");

    // Homogenize line endings from CRLF to LF to make sure tests work both on Windows and Linux/macOS
    let contents = String::from_utf8(contents)
        .unwrap()
        .replace("\r\n", "\n")
        .as_bytes()
        .to_owned();

    let span_offset = compiler.span_offset();
    compiler.add_file(&fname.to_string_lossy(), &contents);

    let parser = Parser::new(compiler, span_offset);
    compiler = parser.parse();

    let mut result = compiler.display_state();

    if !compiler.errors.is_empty() {
        return result;
    }

    let mut resolver = Resolver::new(&compiler);
    resolver.resolve();
    result.push_str(&resolver.display_state());

    compiler.merge_name_bindings(resolver.to_name_bindings());

    result
}

#[test]
fn test_node_output() {
    insta::glob!("../tests", "*.nu", |path| {
        insta::assert_snapshot!(evaluate_example(path));
    });
}
