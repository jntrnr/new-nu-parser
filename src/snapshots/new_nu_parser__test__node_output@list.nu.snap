---
source: src/test.rs
expression: evaluate_example(path)
input_file: tests/list.nu
---
==== COMPILER ====
0: Int (1 to 2)
1: Int (4 to 5)
2: Int (7 to 8)
3: List([NodeId(0), NodeId(1), NodeId(2)]) (0 to 8)
4: Block(BlockId(0)) (0 to 9)
==== SCOPE ====
0: Frame Scope, variables: [  ], node_id: NodeId(4)
==== TYPES ====
0: unknown
1: unknown
2: unknown
3: unknown
4: ()
==== TYPE ERRORS ====
Error (NodeId 3): unsupported ast node 'List([NodeId(0), NodeId(1), NodeId(2)])' in typechecker

